crates/red_knot/src/module.rs

```diff
@@ -0,0 +1,708 @@
+        } else {
+            // This path is for a module with the same name but in a module search path with a lower priority.
+            // Ignore it.
+            None
+        }
+    }
+
+    pub fn add_module(&mut self, path: &Path) {
+        self.resolve_path(path);
+    }
+
+    pub fn remove_module(&mut self, id: ModuleId) {
+        match self.modules.entry(id) {
+            Entry::Occupied(entry) => {
+                // Keep the entry to ensure it locks any writes to the same entry in `resolve`.
+
+                self.by_name.remove(&entry.get().name);
+                // This isn't fast but removing seems an uncommon operation where I don't think it's worth
+                // to keep a reverse lookup just for it (requires more allocations, involves writing a lot of data).
+                self.by_path.retain(|_, current_id| *current_id != id);
+
+                entry.remove();
+            }
+            Entry::Vacant(_) => {
+                panic!("Expected entry to exist")
+            }
+        }
+    }
+}
+
+impl std::fmt::Debug for ModuleResolver {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ModuleResolver")
+            .field("search_paths", &self.search_paths)
+            .field("modules", &self.by_name)
+            .finish()
+    }
+}
+
+fn resolve_name(
+    name: &ModuleName,
+    search_paths: &[ModuleSearchPath],
+) -> Option<(ModuleSearchPath, PathBuf)> {
+    'search_path: for search_path in search_paths {
+        let mut package_path = search_path.path().to_path_buf();
+
+        let mut components = name.components();
+        let module_name = components.next_back()?;
+
+        // For `foo.bar.baz`, test that `foo` and `baz` both contain a `__init__.py`.
+        for folder in components {
+            package_path.push(folder);
+
+            // A folder on the path to the module. Test that it contains a `__init__.py`.
+            if !package_path.join("__init__.py").is_file()
+                && !package_path.join("__init__.pyi").is_file()
+            {
+                // Try the next search path
+                continue 'search_path;
+            }
+        }
+
+        package_path.push(module_name);
+
+        // Must be a `__init__.pyi` or `__init__.py` or it isn't a package.
+        if package_path.is_dir() {
+            package_path.push("__init__");
+        }
+
+        let stub = package_path.with_extension("pyi");
+
+        if stub.is_file() {
+            return Some((search_path.clone(), stub));
+        }
+
+        let module = package_path.with_extension("py");
+
+        if module.is_file() {
+            return Some((search_path.clone(), module));
+        }
+    }
+
+    None
+}
+
+/// The resolved path of a module.
+///
+/// It should be highly likely that the file still exists when accessing but it isn't 100% guaranteed
+/// because the file could have been deleted between resolving the module name and accessing it.
```

Comment:
It would result in us eagerly loading files even if they're unchanged. 

Let's say we load a persistent cache and only have to recheck a single file. The deserialization layer of the persistent cache must acquire module ids for each cached file (and potentially remap them in case it gets assigned a new module id). What we want to avoid is to load the file content at this stage (because it would be very expensive). 

Overall, I think there are a handful operations in the analysis where we just need to know to which module an import resolves but we don't want to analyze the file's content yet (and may not have to because it's either a local analysis or we determine that we have all data cached)

Reply:
Another reason for storing them separately is that we probably want to implement some garbage collection mechanism for source text and AST but I don't think it is necessary for modules, because the data that we store is relatively small. 

The idea for source text and the AST is that we can evict files that haven't been from the cache. For example, we may need to load all files for the initial checking but we can evict most data from the cache (and only keeping the type level cache) once the initialization is over.

Reply:
Re the first comment, I'm not totally clear which are the cases where we would cache the content of a file but not also cache the resolution from module name to that file.

The idea that we may end up having module information without the source text due to cache eviction makes sense.